Maple2.Server.Game/PacketHandlers/UgcHandler.cs

```
@@ -61,25 +67,97 @@ private enum Command : byte {
+            if (voucher != null) {
+                if (!session.Item.Inventory.Consume(voucher.Uid, 1)) {
+                    session.Send(NoticePacket.Notice(NoticePacket.Flags.Alert | NoticePacket.Flags.Message, StringCode.s_err_invalid_item));
+                    return;
+                }
+            }
+        } else {
+            switch (ugcMetadata.CurrencyType) {
+                case MeretMarketCurrencyType.Meso:
+                    if (session.Currency.CanAddMeso(-ugcMetadata.CreatePrice) != -ugcMetadata.CreatePrice) {
+                        session.Send(NoticePacket.MessageBox(StringCode.s_err_lack_meso));
+                        return;
+                    }
+                    session.Currency.Meso -= ugcMetadata.CreatePrice;
+                    break;
+                case MeretMarketCurrencyType.Meret:
+                    if (session.Currency.CanAddMeret(-ugcMetadata.CreatePrice) != -ugcMetadata.CreatePrice) {
+                        session.Send(NoticePacket.MessageBox(StringCode.s_err_lack_merat));
+                        return;
+                    }
+                    session.Currency.Meret -= ugcMetadata.CreatePrice;
+                    break;
+                case MeretMarketCurrencyType.RedMeret:
+                    if (session.Currency.CanAddGameMeret(-ugcMetadata.CreatePrice) != -ugcMetadata.CreatePrice) {
+                        session.Send(NoticePacket.MessageBox(StringCode.s_err_lack_merat_red));
+                    }
+                    session.Currency.GameMeret -= ugcMetadata.CreatePrice;
+                    break;
+                default:
+                    Logger.Error("Unhandled currency type {UgcMetadataCurrencyType}", ugcMetadata.CurrencyType);
+                    return;
+            }
+        }
+
+        Item? item = session.Item.CreateItem(itemId, ugcMetadata.ItemRarity);
+        if (item == null) {
+            return;
```

This needs to be logged as fatal since the player already paid for this we shouldn't silently fail.

something similar to:
```cs
Log.Fatal("Failed to consume ingredient {ItemUid}", ingredient.Uid);
throw new InvalidOperationException($"Fatal: Consuming ingredient: {ingredient.Uid}");
```